openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/edgefilters/core/ExclusiveDownwardSearchEdgeFilter.java

```diff
@@ -0,0 +1,45 @@
+/*  This file is part of Openrouteservice.
+ *
+ *  Openrouteservice is free software; you can redistribute it and/or modify it under the terms of the
+ *  GNU Lesser General Public License as published by the Free Software Foundation; either version 2.1
+ *  of the License, or (at your option) any later version.
+
+ *  This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *  See the GNU Lesser General Public License for more details.
+
+ *  You should have received a copy of the GNU Lesser General Public License along with this library;
+ *  if not, see <https://www.gnu.org/licenses/>.
+ */
+package org.heigit.ors.routing.graphhopper.extensions.edgefilters.core;
+
+import com.graphhopper.routing.profiles.BooleanEncodedValue;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.storage.CHGraph;
+import com.graphhopper.util.EdgeIteratorState;
+import org.heigit.ors.routing.graphhopper.extensions.edgefilters.ch.CHLevelEdgeFilter;
+
+public class ExclusiveDownwardSearchEdgeFilter extends CHLevelEdgeFilter {
+	protected final BooleanEncodedValue accessEnc;
+	private boolean swap = false;
+
+	public ExclusiveDownwardSearchEdgeFilter(CHGraph g, FlagEncoder encoder) {
+		super(g, encoder);
+		accessEnc = encoder.getAccessEnc();
+	}
+
+
+	public ExclusiveDownwardSearchEdgeFilter(CHGraph g, FlagEncoder encoder, boolean swap) {
+		this(g, encoder);
+		this.swap = swap;
+	}
+
+	@Override
+	public boolean accept(EdgeIteratorState edgeIterState) {
+		int adj = edgeIterState.getAdjNode();
+		if (baseNode >= maxNodes || adj >= maxNodes || baseNodeLevel < graph.getLevel(adj))
```

Comment:
As you pointed out before, `ExclusiveDownwardSearchEdgeFilter` differs from `DownwardSearchEdgeFilter` only by the different condition `baseNodeLevel < graph.getLevel(adj)` vs. `baseNodeLevel <= graph.getLevel(adj)`, respectively. I think things could be cleaned up even more by:

1. making `ExclusiveDownwardSearchEdgeFilter` a subclass of `DownwardSearchEdgeFilter`
2. delegating the check `baseNodeLevel <= graph.getLevel(adj)` in `DownwardSearchEdgeFilter` to a method, which could be overridden in `ExclusiveDownwardSearchEdgeFilter`. 

What do you think?

Reply:
Oops, missed the `swap` field. Still, it makes sense for `ExclusiveDownwardSearchEdgeFilter` to be a subclass of `DownwardSearchEdgeFilter` as it would eliminate some redundancy.

Reply:
I don't think so. < and <= are different relations. None is more prominent to be a subclass of the other.

Reply:
I also think this would overcomplicate things. To do it properly, you'd have to write some Abstract Downwardfilter and then implement an exclusive and an inclusive version. I think this is overkill for a single `=`

Reply:
Thanks @sfendrich for chiming in! While in principle `<` are `<=` different relations, they are only used for implementing the broader functionality. The practical difference between these two filters is, as indicated by the class name, that `ExclusiveDownwardSearchEdgeFilter` accepts a subset of edges which `DownwardSearchEdgeFilter` does. In this context it doesn't seem wrong to me to think of the exclusive filter as a specialization, hence subclass of the broader inclusive one. Both filters share `accessEnc` which I consider as unnecessary code duplication which could be eliminated by the suggested class hierarchy. Note that even the current naming of the filters already suggest this type of relation between the classes.

Reply:
I changed it, but still I don't think this is worth it in any way. I prefer the very first implementation that I had where it was a branched `accept` with the exclusivity of the sign being a simple `boolean`. That is the least additional code and I find it quite clear. Not planning on reverting it back, though.

Reply:
Thanks a lot! I got your point, it's easier to follow if everything is one place. Nevertheless from performance perspective it's generally better to choose solutions which avoid branching. You know it for sure, just in case you are interested is some hardcore details I can recommend [this article](https://johnysswlab.com/how-branches-influence-the-performance-of-your-code-and-what-can-you-do-about-it/).

Reply:
Thanks. I am aware of branching issues. I am also aware that Java uses branch prediction, which is trivial in this exact case where there is a boolean that can only be set in the constructor.
Anyway, this is all nice and great theoretical talk, but I want this to get done. Do you accept the solution where the exclusive filter extends the normal one? Then we can move an.

Reply:
Sure, thanks again for following up on my suggestion!

Reply:
Still, I think making one a subclass of the other is not a good idea. The primary reason to use subclasses is to enable dynamic polymorphism in the sense of the Liskov Substitution Principle (LSP). Real world conceptual hierarchies (like one being a subset of the other) do not map well to OO inheritance hierarchies; see the well-known circle-ellipse or square-rectangle problems. Having to override a non-abstract method already indicates the risk of violating the LSP and, hence, of hurting maintainability with bugs that are hard to find.
Possible solutions to this issue are, e.g.,
* Introducing an interface above both
* @HendrikLeuschner 's proposal of using an abstract class instead
* An abstract class combined with the template method pattern
* My preferred solution: injecting the comparator as strategy in the strategy pattern, e.g., using lambda-notation

Reply:
Also note, that you don't avoid branching, you just move it into a different place, namely, the compiler-generated code that chooses the right method to call.

Reply:
Which one is it going to be then? Please, goal-driven work. I will reimplement once more, but that is the last time.

Reply:
If the deadline is an issue, then leave it as is, and add a TODO comment in the code.

Reply:
Thanks @sfendrich for suggesting some alternative implementations! I didn't intend to spark a huge discussion here. My primary goal was to introduce some DRYness for better code maintainability and an attempt to avoid obvious branching. For the sake of the latter a compile-time solution, so the template-based one, is probably the preferred approach. Note that the difference between the filters is not only about the operators being different. There is some additional logic based on the `swap` field specific to the `Exclusive...` class so I'm not sure the idea based on the strategy pattern would work well here, what do you think?

> Also note, that you don't avoid branching, you just move it to a different place, namely, the compiler-generated code that chooses the right method to call.

Fair point! TBH I didn't really think about this before. I was naively expecting that at run time once an object of a certain class has been created there is a some smart way to avoid the branching each time the object's method is called. Could you maybe point me to some further reading on this topic? I would love to learn how polymorphic method calls actually translate into bytecode.

Reply:
Concerning branch prediction, please see my edit above.
The exact translation is implementation dependent. However, the JVM spec gives quite some details: https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-6.html#jvms-6.5.invokevirtual.